intellij-ext/modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordConfigDropDown.kt

```diff
@@ -111,16 +111,46 @@ class MirrordConfigDropDown : ComboBoxAction() {
         configFiles = updatedConfigFiles
     }
 
-    private fun getReadablePath(path: String, project: Project): String {
-        val basePath = project.basePath ?: throw Error("couldn't resolve project path")
-        val relativePath = Path.of(basePath).relativize(Path.of(path))
-        return relativePath.toString()
+    // this function spawns a background task to query the index, not blocking the current thread
+    // it maintains a single task to do so, using the `blockQueries` flag
+    private fun queryIndexInSmartMode(project: Project, query: (project: Project) -> Unit) {
+        if (!blockQueries.getAndSet(true)) {
+            object : Task.Backgroundable(project, "mirrord") {
+                override fun run(indicator: ProgressIndicator) {
+                    val dumbService = DumbService.getInstance(project)
+                    // refer to `DumbService`, there is no "guarantee" still, but we stay in a loop
+                    // todo: should probably look into using the message bus to listen for indexing to finish
+                    while (true) {
+                        dumbService.waitForSmartMode()
+                        val success = ReadAction.compute<Boolean, RuntimeException> {
+                            if (project.isDisposed) {
+                                throw ProcessCanceledException()
+                            }
+                            if (dumbService.isDumb) {
+                                return@compute false
+                            }
+                            try {
+                                indicator.text = "mirrord: updating config files"
+                                query(project)
+                                blockQueries.set(false)
```

Comment:
This synchronisation mechanism is incorrect. Possible scenario:
1. In background task 1 (after file system change): `query(project)` (`updateConfigFiles(project)`) completes.
2. New filesystem change, so in updating thread: `queryIndexInSmartMode` is called, `blockQueries` is still true, so not creating new task.
3. In background task 1: `blockQueries.set(false)` (too late).
Now as long as there is no new file system change, `queryIndexInSmartMode` (`updateConfigFiles`) is not called, and the last change is not visible for the user.

Reply:
> but the update flag is still set to false

When the update flag is set to false, the update will not happen in the next call to `update`.

> not even need atomics
About the atomics already in `main`, see https://github.com/metalbear-co/mirrord/pull/1395#discussion_r1190325799. About the atomics added in this PR, I don't know, please explain why this PR is needed.